desktop/src/components/Terminal/useTerminalSearch.tsx

```diff
@@ -0,0 +1,402 @@
+}
+
+type THighlight = {
+  displayRow: number
+
+  // Bounds of the highlight within a row.
+  startCol: number
+  endCol: number
+
+  // Which search result this highlight is attached to.
+  resultIndex: number
+}
+
+type TDisplayLineMap = { [key: number]: TDisplayLine[] }
+
+// Used for keeping track of anchors/jump points in the wrapped lines to jump to for any given search result.
+type TJumpMap = { [resultIndex: number]: number }
+
+type TSearchState = {
+  searchableLines: string[]
+  disposables: VoidFunction[]
+  searchOptions: TSearchOptions | undefined
+  jumpMap: TJumpMap
+  debounce: number | undefined
+  activeSearchResult: number
+  preWrappedLines: TDisplayLineMap | undefined
+  searchResults: TSearchResult[]
+  highlights: THighlight[]
+}
+
+export function useTerminalSearch(
+  terminalRef: React.MutableRefObject<TTerminal | null>,
+  searchOptions?: TSearchOptions
+) {
+  const [totalSearchResults, setTotalSearchResults] = useState<number>(-1)
+  const [activeSearchResult, setActiveSearchResult] = useState<number>(0)
+
+  // We have to exercise caution not to re-generate connectStream,
+  // so we have to store a lot of state outside the usual mechanisms.
+  // Otherwise, we will make the terminal flicker.
+
+  const searchStateRef = useRef<TSearchState>({
+    searchableLines: [],
+    disposables: [],
+    searchOptions,
+    jumpMap: {},
+    debounce: undefined,
+    activeSearchResult: 0,
+    preWrappedLines: undefined,
+    searchResults: [],
+    highlights: [],
+  })
+
+  const clearDisposables = useCallback(() => {
+    const toClear = searchStateRef.current.disposables
+    searchStateRef.current.disposables = []
+    toClear.forEach((disposable) => disposable())
+  }, [searchStateRef])
+
+  const repaintHighlights = useCallback(
+    (highlights: THighlight[]) => {
+      const terminal = terminalRef.current
+      const displayLines = searchStateRef.current.preWrappedLines
+      const searchResults = searchStateRef.current.searchResults
+
+      if (!displayLines || !terminal || !searchResults.length) {
+        return
+      }
+
+      searchStateRef.current.highlights = highlights
+
+      for (const highlight of highlights) {
+        const isActive = highlight.resultIndex === searchStateRef.current.activeSearchResult
+
+        const disposable = terminal.highlight(
+          highlight.displayRow,
+          highlight.startCol,
+          highlight.endCol - highlight.startCol,
+          isActive ? "#E4ADFF" : "#8E00EB",
+          isActive
+        )
+
+        if (disposable) {
+          searchStateRef.current.disposables.push(disposable)
+        }
+      }
+    },
+    [searchStateRef, terminalRef]
+  )
+
+  // When the terminal is resized, we need to re-calculate the highlights and jump anchors,
+  // as these have to be positioned per wrapped line.
+  const onResize = useCallback(
+    (cols: number) => {
+      const displayLines = (searchStateRef.current.preWrappedLines = wrapLines(
```

Comment:
nit: could you extract this into two lines please?

Reply:
Done